packages/core/src/magicswap/swap.ts

```diff
@@ -0,0 +1,196 @@
+import { type Config, connect, mock, readContract } from "@wagmi/core";
+import { erc20Abi } from "viem";
+import { erc1155Abi } from "../abis/erc1155Abi";
+import type { AddressString, SupportedChainId } from "../types";
+import { getContractAddresses } from "../utils/contracts";
+import { DEFAULT_WAGMI_CONFIG } from "../utils/wagmi";
+import type { Pool } from "./fetchPools";
+import { getSwapRoute } from "./getSwapRoute";
+import type { PoolToken } from "./types";
+
+type NFTInput = { id: string; quantity: number };
+
+const getAmountMax = (amount: bigint, slippage: number) =>
+  amount + (amount * BigInt(Math.ceil(slippage * 1000))) / 1000n;
+const getAmountMin = (amount: bigint, slippage: number) =>
+  amount - (amount * BigInt(Math.ceil(slippage * 1000))) / 1000n;
+
+const DEFAULT_SLIPPAGE = 0.005;
+
+export const swap = async ({
+  pools,
+  toAddress,
+  tokenInId,
+  tokenOutId,
+  nftsIn,
+  nftsOut,
+  amountIn = "0",
+  amountOut = "0",
+  isExactOut,
+  chainId,
+  slippage = DEFAULT_SLIPPAGE,
+  wagmiConfig = DEFAULT_WAGMI_CONFIG,
+}: {
+  pools: Pool[];
+  toAddress: AddressString;
+  tokenInId: string;
+  tokenOutId: string;
+  nftsIn: NFTInput[];
+  nftsOut: NFTInput[];
+  amountIn?: string;
+  amountOut?: string;
+  isExactOut: boolean;
+  chainId: SupportedChainId;
+  slippage?: number;
+  wagmiConfig?: Config;
+}) => {
+  await connect(wagmiConfig, {
+    chainId,
+    connector: mock({
+      accounts: ["0x8c59E81e2553104d0B4F3dE19D0eD347055BB218"],
+    }),
+  });
+
+  const poolTokens = pools
+    .flatMap(({ token0, token1 }) => [token0, token1])
+    .reduce(
+      (acc, poolToken) => {
+        acc[poolToken.id] = poolToken;
+        return acc;
+      },
+      {} as Record<string, PoolToken>,
+    );
+
+  const tokenIn = poolTokens[tokenInId];
+  const tokenOut = poolTokens[tokenOutId];
+  const contractAddresses = getContractAddresses(chainId);
+  const magicSwapV2RouterAddress = contractAddresses.MagicSwapV2Router;
+  const deadline = BigInt(Math.floor(Date.now() / 1000) + 30 * 60);
+
+  // From NFT
+  if (tokenIn.isNFT) {
+    // To NFT
+    if (tokenOut.isNFT) throw new Error("NFT to NFT swaps are not supported");
+
+    // To Token
+    const collectionId = tokenIn.collectionId;
+    const amountOutMin = isExactOut
+      ? BigInt(amountOut)
+      : getAmountMin(BigInt(amountOut), slippage);
+
+    const route = getSwapRoute({
+      pools,
+      tokenInId,
+      tokenOutId,
+      amount: amountOut,
+      isExactOut,
+    });
```

Comment:
We don't need to calculate the route in this function - let the consumer pass the `path` in themselves (see `useSwap` hook in `magicswapv2` repo)